crates/iota-types/src/stardust/address_swap_map.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, str::FromStr};
+
+use serde::Deserialize;
+
+use crate::base_types::IotaAddress;
+type OriginAddress = IotaAddress;
+type DestinationAddress = IotaAddress;
+
+#[derive(Debug, Default, Deserialize)]
+pub struct AddressSwapMap {
+    addresses: HashMap<OriginAddress, DestinationAddress>,
+    address_swapped_at_least_once: HashMap<OriginAddress, bool>,
+}
+
+impl AddressSwapMap {
+    /// Retrieves the destination address for a given origin address.
+    /// Marks the origin address as swapped if found.
+    pub fn get_destination_address(
+        &mut self,
+        origin_address: OriginAddress,
+    ) -> Option<&DestinationAddress> {
+        self.addresses.get(&origin_address).inspect(|_| {
+            // Mark the origin address as swapped
+            if let Some(swapped_flag) = self.address_swapped_at_least_once.get_mut(&origin_address)
+            {
+                *swapped_flag = true;
+            }
+        })
+    }
+
+    /// Verifies that all addresses have been swapped at least once.
+    /// Returns an error if any address is not swapped.
+    pub fn verify_all_addresses_swapped(&self) -> anyhow::Result<()> {
+        if let Some((addr, _)) = self
+            .address_swapped_at_least_once
+            .iter()
+            .find(|(_, &is_swapped)| !is_swapped)
+        {
+            return Err(anyhow::anyhow!("address to swap not found: {:?}", addr));
+        }
+        Ok(())
+    }
+}
+
+/// Initializes an [`AddressSwapMap`] by reading address pairs from a CSV file.
+///
+/// The function expects the file to contain two columns: the origin address
+/// (first column) and the destination address (second column). These are parsed
+/// into a [`HashMap`] that maps origin addresses to destination addresses.
```

Comment:
I was trying the change locally and would like to add some comments:
 - The format of the csv file is not well defined, one can use whatever headers. Would be nice to force certain headers (eg. `from,to`) and check them in `init_address_swap`.
 - It's unclear to the user that the `from` addresses of the csv file shall be in hex encoded new address format. It is actually impossible to derive this without code for the average user. Therefore, I would like to propose to expect the `from` address in the old `bech32` format.
 - Adding an example `csv` file either in a README, cli help, or code comments would help the usage of the application tremendously, as it would clearly show what format is expected.

Reply:
1. I can add headers like Origin,Destination
2. I didn’t expect we’d have the option for custom CSV file support—I thought we’d use it once and get it from the DevOps team.
3. I’ll add an example

Reply:
3. Examples have been added here https://github.com/iotaledger/iota/pull/4389 and here https://github.com/iotaledger/iota/blob/d21f2b501a0b239e162865eca02afb28081e0e94/crates/iota-genesis-builder/src/stardust/types/address_swap_map.rs#L137